### PR TITLE
Add “Component - Styling” content

### PR DESCRIPTION
**Pull request by knowler**
Making some actual progress with the prose for the article. Sections can be moved around. It might even be good to break this up into separate article since it’s a big topic. If so, I can expand the intro to be more of an overview which mentions each of them.

### CODE REVIEW

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:76]: I think this paragraph needs some refinement. I feel uneasy about "couple types of elements". Perhaps splitting the two concepts would work better... something like:

```suggestion
In order to be portable, Web Components can provide default styles for the element itself (also known as the shadow host).
They can also style slotted elements with some default styles.
```

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:82]: Front-loading knowledge of both selector types here is really nice :ok_hand:

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:97]: This paragraph feels a bit too contingent on surrounding context. Perhaps:

```suggestion
The `:host()` selector will select the shadow host if it matches a given selector. For example, it can be  used to
select for hosts that have a given attribute. While `:host` may apply to `<fancy-p>` component, `:host([extra])` would
apply only to `<fancy-p extra>` elements:
```

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:130]: This feels like it's not relevant to the heading here. Maybe this can be removed?

```suggestion
```

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:135]: "different kinds of elements" feels a little ambiguous to me. I wonder if this works:

```suggestion
The `::slotted()` pseudo-element selector allows you to write default styles for slotted elements that match the given
selector. Specifying a selector can be useful for styling specific elements in particular ways.
```

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:187]: This feels important enough to have a subheading, I feel. Also, I wonder if we can present alternative options. How does one overcome this barrier (more custom elements :smiley:)?

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:132]: One thing missing from this section is `::slotted(*)` which feels like something worth documenting.

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:195]: I definitely think Parts deserve their own section, especially as by this point I think we're close to 10 minutes of reading time.

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:243]: It might be worth lifting this section way up to the top of the document.

File: learn/components/styling.md
Comment (keithamus) [inline learn/components/styling.md:272]: This would need to be in `<head>` right?

**Review by keithamus — APPROVED** [review]
Phil Hartman spitting out a cigarette in amazement, and saying "wow"


This is looking really really great so far! I've added a couple of notes which include some questions and some edits I think might tighten up a couple of pieces, but this is in super great shape already; I think we could merge this as is and continue iterating on it, if you'd like!

File: learn/components/styling.md
Comment (knowler) [reply learn/components/styling.md:15]: I think that’s a great suggestion. I definitely agree with the hard-to-readness of the first two sentences despite their careful wording and symmetry.

File: learn/components/styling.md
Comment (knowler) [reply learn/components/styling.md:271]: Not necessary, technically it could be anywhere since it isn’t rendered, but on page load `<head>` would be the best place. A scenario where it wouldn’t really matter if it were in the `<head>` would be if a web component was dynamically imported later on. That’s a more advanced use case though. The thing with this suggestion in general is that the preload and the usage of the component on the page can’t happen at different times, otherwise, the preload is worthless and you’ll get a console warning about it. I think maybe I can rework this suggestion to emphasize the caution in doing this. Generally I think `<style>` and constructable stylesheets seem better because of this.

File: learn/components/styling.md
Comment (knowler) [reply learn/components/styling.md:194]: Here’s a thought I just had for how to structure all the styling content maybe in different section:

- Introduction
  - Focuses on encapsulation and then gives a brief overview with links to each of the following sections.

- Ways to include styles in the Shadow DOM
  - A necessary but boring prerequisite for everything.

- How to write default styles for Web Components.
  - Covers `:host` and `::slotted`.
  - Shows how to make Web Component Styling portable.

- How to create a styling API for Web Components.
  - Covers using custom properties, `::part()`, `exportparts` for nested custom elements, and maybe even `!important` to limit styling certain properties.
  - Shows how to make Web Component Styling extensible.

There is more to cover like gotchas around `@keyframes` and `@font-face`. Maybe those could go in the “default styles” section.

**Comment by knowler** [comment]
@keithamus That’s for the review. I applied your suggestions since I thought they were great. I think merging it is a good idea. Having it live will probably help drive me to keep working on it. Also, take a look at this comment I left when you get a chance (it isn’t blocking merging though): https://github.com/WebComponentsGuide/webcomponents.guide/pull/66#discussion_r1068636643

File: learn/components/styling.md
Comment (keithamus) [reply learn/components/styling.md:194]: I really like it. I like the idea of layering knowledge. What's the first thing that someone needs to just get styles working? Armed with that knowledge up front, they can experiment, and learn things independently, rather than having to read the entirety. Perhaps the basics can be expanded upon. Talk more about how it could be preferable to simplify selectors given the encapsulation. Maybe the introduction and "ways to include..." can be combined into a "Here's the basics" section.

Parts, while foundational, are more about architecting a component. So we should batch that with other architectural parts. So yeah I think having a section about styling API & extensibility is great.

**Comment by keithamus** [comment]
I'll leave the joy of merging this to you 😃

**Comment by knowler** [comment]
Just realized I should have fixed the failing tests before merging. Will have to follow up in a bit.

**Pull request merged**
